test/bike_brigade/importers/mailchimp_importer_test.exs

```diff
@@ -15,36 +22,78 @@ defmodule BikeBrigade.Importers.MailchimpImporterTest do
       TEXTYUI_3: "1508 Davenport Rd"
     }
   }
-  test "parse_mailchimp_attrs is successful if fields from Mailchimp are valid" do
-    {:ok, expected_fields} = MailchimpImporter.parse_mailchimp_attrs(@valid_attrs)
-    address = expected_fields[:location_struct][:address]
-    assert address == "1508 Davenport Rd Toronto"
-  end
 
-  test "parse_mailchimp_attrs returns an error if the phone number is not a valid Canadian number" do
-    invalid_attrs = put_in(@valid_attrs, [:merge_fields, :PHONEYUI_], "oops")
-    {:error, _} = MailchimpImporter.parse_mailchimp_attrs(invalid_attrs)
-  end
+  describe "parse_mailchimp_attrs/1" do
+    test "is successful if fields from Mailchimp are valid" do
+      {:ok, expected_fields} = MailchimpImporter.parse_mailchimp_attrs(@valid_attrs)
+      address = expected_fields[:location_struct][:address]
+      assert address == "1508 Davenport Rd Toronto"
+    end
 
-  test "parse_mailchimp_attrs returns an error if address fails geocoding" do
-    invalid_attrs = put_in(@valid_attrs, [:merge_fields, :TEXTYUI_3], "oops")
+    test "returns an error if the phone number is not a valid Canadian number" do
+      invalid_attrs = put_in(@valid_attrs, [:merge_fields, :PHONEYUI_], "oops")
+      {:error, _} = MailchimpImporter.parse_mailchimp_attrs(invalid_attrs)
+    end
 
-    {:error, {:update_location, _}} = MailchimpImporter.parse_mailchimp_attrs(invalid_attrs)
-  end
+    test "returns an error if address fails geocoding" do
+      invalid_attrs = put_in(@valid_attrs, [:merge_fields, :TEXTYUI_3], "oops")
 
-  test "sync_riders is successful when all required attributes are valid" do
-    {:ok, _} = MailchimpImporter.sync_riders({:ok, [@valid_attrs]})
+      {:error, {:update_location, _}} = MailchimpImporter.parse_mailchimp_attrs(invalid_attrs)
+    end
   end
 
-  test "sync_riders handles case where phone is invalid" do
-    invalid_attrs = put_in(@valid_attrs, [:merge_fields, :TEXTYUI_3], "oops")
+  describe "sync_riders/0" do
+    setup do
+      saved_env = Application.get_env(:bike_brigade, BikeBrigade.Importers.MailchimpImporter)
 
-    {:ok, _} = MailchimpImporter.sync_riders({:ok, [invalid_attrs]})
-  end
+      Application.put_env(
+        :bike_brigade,
+        BikeBrigade.Importers.MailchimpImporter,
+        Keyword.put(saved_env, :list_id, @list_id)
```

Comment:
Interesting! Is this updating some kind of global state? Any links to elixir testing info I can read more about?

Reply:
Yup  -- the environment is a global key/value state https://hexdocs.pm/elixir/Application.html#module-the-application-environment

The stuff in config.exs and dev.exs/test.exs/prod.exs ends up being put there at compile time, and the stuff in runtime.exs gets there at runtime.

I could have put a list id in test.exs, but I felt that it didn't belong there since it was an arbitrary string used in mocking our importer and not an actual testing configuration thing.

The docs for ExUnit.Case are a good intro to testing: https://hexdocs.pm/ex_unit/1.12/ExUnit.Case.html (I generally find the elixir docs excellent but you need to know which module name you're looking for to get to the good stuff :( )

It might be useful to look at the changes I made in your tests: https://github.com/bikebrigade/dispatch/pull/50/files/96b3a0bda0498eb5d738369d27d90504c7ae1f10..4c38173f6a33492cf6ccac080862b5045de65a15#diff-15a82d00038d613b51a9716bfafd946bbe606c7e3b3018f3c287c2cc5cfc8de0

I think the main things I introduced were
1) `assert` works with pattern matching as well, e.g. `assert {:ok, _} = foo()` -- just putting the pattern match without the assert will still error and fail the test, but the assert will give the correct feedback and make it a test failure instead of an exception.
2) Added some `describe`s, this is pretty self explanatory but note you only have one level of `describe`s. Jose talks about some of the thinking on that (and being stubborn about adding it at all) on this excellent recent podcast episode: https://podcast.thinkingelixir.com/82
3) `setup/on_exit` to setup the application environment. See https://hexdocs.pm/ex_unit/1.12/ExUnit.Callbacks.html#on_exit/2 and that module's doc for the test lifecycle.

I didn't use them, but [test contexts](https://hexdocs.pm/ex_unit/1.12/ExUnit.Callbacks.html) are a little counterintuitive and used in a lot of other tests.

If you want to go deeper, note that the test starts with `use BikeBrigade.DataCase`. This is a module that lives in our project - it's been autogenerated for us by phoenix. Worth taking a look: https://github.com/bikebrigade/dispatch/blob/main/test/support/data_case.ex

If you want to go *really* deep, the [code for `assert`](https://github.com/elixir-lang/elixir/blob/a64d42f5d3cb6c32752af9d3312897e8cd5bb7ec/lib/ex_unit/lib/ex_unit/assertions.ex#L45) is a great example of the power of macros